pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1075,6 +1075,13 @@ private void closeConsumerTasks() {
         }
         negativeAcksTracker.close();
         stats.getStatTimeout().ifPresent(Timeout::cancel);
+        // Execute "clearIncomingMessages" regardless of whether "internalPinnedExecutor" has shutdown or not.
+        // Call "clearIncomingMessages" in "internalPinnedExecutor" is used to clear messages in flight.
+        if (!internalPinnedExecutor.isShutdown()) {
+            internalPinnedExecutor.execute(this::clearIncomingMessages);
+        } else {
+            clearIncomingMessages();
```

Comment:
If the `executor` has shutdown, it is still running some task in background.

Reply:
fixed